src/credentials/activation.ts

```diff
@@ -40,36 +42,49 @@ export async function loginWithMostRecentCredentials(
     const profileNames = Object.keys(providerMap)
     const previousCredentialsId = toolkitSettings.readSetting<string>(profileSettingKey, '')
 
+    async function tryConnect(creds: CredentialsProviderId, popup: boolean): Promise<boolean> {
+        const provider = await manager.getCredentialsProvider(creds)
+        // 'provider' may be undefined if the last-used credentials no longer exists.
+        if (!provider) {
+            getLogger().warn('autoconnect: getCredentialsProvider() lookup failed for profile: %O', asString(creds))
+        } else if (provider.canAutoConnect() && (await loginManager.login({ passive: true, providerId: creds }))) {
+            getLogger().info('autoconnect: connected with profile: %O', asString(creds))
+            if (popup) {
+                vscode.window.showInformationMessage(
+                    localize('AWS.message.credentials.connected', 'Connected to AWS as {0}', asString(creds))
+                )
+            }
+            return true
+        }
+        return false
+    }
+
+    if (!previousCredentialsId && !(providerMap && profileNames.length === 1)) {
+        getLogger().info('autoconnect: skipped')
+        return
+    }
+
     if (previousCredentialsId) {
         // Migrate from older Toolkits - If the last providerId isn't in the new CredentialProviderId format,
         // treat it like a Shared Crdentials Provider.
         const loginCredentialsId = tryMakeCredentialsProviderId(previousCredentialsId) ?? {
             credentialType: SharedCredentialsProvider.getCredentialsType(),
             credentialTypeId: previousCredentialsId,
         }
-        const provider = await manager.getCredentialsProvider(loginCredentialsId)
-
-        // 'provider' may be undefined if the last-used credentials no longer exists.
-        if (provider && provider.canAutoConnect()) {
-            await loginManager.login({ passive: true, providerId: loginCredentialsId })
-        } else {
-            await loginManager.logout()
+        if (await tryConnect(loginCredentialsId, false)) {
+            return
         }
-    } else if (
-        providerMap &&
-        profileNames.length === 1 &&
-        (await manager.getCredentialsProvider(providerMap[profileNames[0]]))!.canAutoConnect()
-    ) {
+    }
+
+    if (providerMap && profileNames.length === 1) {
         // Auto-connect if there is exactly one profile.
-        if (await loginManager.login({ passive: true, providerId: providerMap[profileNames[0]] })) {
-            // Toast.
-            vscode.window.showInformationMessage(
-                localize('AWS.message.credentials.connected', 'Connected to AWS as {0}', profileNames[0])
-            )
+        if (await tryConnect(providerMap[profileNames[0]], !isCloud9())) {
```

Comment:
Is the popup valuable if isCloud9() but the user set a different profile name?

Reply:
I don't know. I chose to avoid popups aggressively on cloud9, since in the common case the popup is redundant/undesired. Adding extra logic is an unnecessary risk for now.